### PR TITLE
LG-4620: `ChartCard` and `Header` components

### PR DESCRIPTION
**Pull request by tsck**
## ✍️ Proposed changes

This adds both `ChartCard` and `Header` components.

🎟 _Jira tickets:_ 
- [LG-4620](https://jira.mongodb.org/browse/LG-4620)
- [LG-4575](https://jira.mongodb.org/browse/LG-4575)

### For bug fixes, new features & breaking changes

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

### For new components

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README
- [ ] I have verified the Live Example looks as intended on the design website.

## 🧪 How to test changes
- See working in Storybook
- Confirm unit tests are passing

### CODE REVIEW

File: charts/core/README.md [hunk 1]
```diff
@@ -58,66 +84,92 @@ Chart container component.
 
 #### Props
 
-| Name           | Description                                             | Type       | Default |
-| -------------- | ------------------------------------------------------- | ---------- | ------- |
-| `onChartReady` | Callback to be called when chart is finished rendering. | () => void |         |
+| Name           | Description                                             | Type         | Default |
+| -------------- | ------------------------------------------------------- | ------------ | ------- |
+| `onChartReady` | Callback to be called when chart is finished rendering. | `() => void` |         |
```
Comment (tsck) [inline charts/core/README.md:89]: Made some drive-by style changes in the README, mostly to use backticks on some code that wasn't previously formatted

File: charts/core/src/core.stories.tsx
Comment (tsck) [inline charts/core/src/core.stories.tsx:1]: I made a bunch of updates to the story to provide more configuration controls. Since there are multiple components shown at once, I thought it made sense to provide controls to turn them all off and on.

File: charts/core/src/ChartCard/ChartCard.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      setHeaderHeight(headerRef.current.offsetHeight);
+    }
+  }, [containerRef, headerRef]);
+
+  return (
+    <div
+      className={cx(
+        getContainerStyles(theme, height, headerHeight),
+        className,
+        isOpen && 'open',
```
Comment (shaneeza) [inline charts/core/src/ChartCard/ChartCard.tsx:70]: We don't typically use this pattern. An alternative could be to pass `isOpen` to `getContainerStyles(theme, height, headerHeight, isOpen)` and then conditionally render the style:

```
export const getContainerStyles = (
  theme: Theme,
  height?: number,
  headerHeight?: number,
  isOpen,
) => cx(css`
  ... default styles here
  `, 
  {
   [css`
       ... open styles here
   `]: isOpen
  }

File: charts/core/src/ChartCard/ChartCard.styles.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  * a wrapper around the children because a bottom border is set on the container and header.
+  * By shrinking the container height we prevent a conflict between the container border 
+  * and the header border when collapsed. By making the container height 1px shorter on collapse
+  * we just cutoff the header border entirely.
+  */
+  /* Accounts for 1px border */
+  max-height: ${headerHeight ? headerHeight + 1 + 'px' : 'auto'};
+  transition: max-height ${transitionDuration.slower}ms ease-in-out;
+  &.open {
+    max-height: ${height ? height + 'px' : 'auto'};
```
[the remark this answers is not in the document]
Comment (tsck) [reply charts/core/src/ChartCard/ChartCard.styles.ts:34]: Oh very interesting, I love this. Will push a change for this shortly!

File: charts/core/src/ChartCard/ChartCard.tsx [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        ref={headerRef}
+      >
+        <div className={leftInnerContainerStyles}>
+          <IconButton
+            className={toggleButtonStyles}
+            id={toggleId}
+            aria-label="Toggle button"
+            aria-controls={childrenId}
+            aria-expanded={isOpen}
+            onClick={(e: MouseEvent<HTMLButtonElement>) => {
```
Comment (shaneeza) [inline charts/core/src/ChartCard/ChartCard.tsx:87]: nit: you could create a function like `handleIconClick` and pass that to `onClick`

File: charts/core/src/ChartCard/ChartCard.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+              className={cx(toggleIconStyles, isOpen && 'open')}
+            />
+          </IconButton>
+          <Body weight="medium" baseFontSize={BaseFontSize.Body2}>
+            {title}
+          </Body>
+        </div>
+        <div>{headerContent}</div>
+      </div>
+      <div style={{ overflow: 'hidden' }}>
```
Comment (shaneeza) [inline charts/core/src/ChartCard/ChartCard.tsx:105]: nit: classname

File: charts/core/src/Header/Header.styles.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import { css } from '@leafygreen-ui/emotion';
+import { Theme } from '@leafygreen-ui/lib';
+import {
+  color,
+  InteractionState,
+  spacing,
+  Variant,
+} from '@leafygreen-ui/tokens';
+
+export const getContainerStyles = (theme: Theme, showDivider?: boolean) => css`
```
Comment (shaneeza) [inline charts/core/src/Header/Header.styles.ts:10]: A tip i got from Adam was to use 
```
cx(css`...`, {[css`...`]:showDivider})
```
instead of 
```
css`
...
${showDivider && ``}
`
```
`cx` is better for conditional classes

File: charts/core/src/Chart/Chart.styles.ts [hunk 7]
```diff
@@ -1,22 +1,16 @@
...
-  InteractionState,
-  Variant,
-} from '@leafygreen-ui/tokens';
 
-export const chartStyles = css`
-  grid-row: 2;
+export const chartContainerStyles = css`
+  display: grid;
+  grid-template-areas:
+    'chartHeader'
```
Comment (shaneeza) [inline charts/core/src/Chart/Chart.styles.ts:6]: nit: Do you explicitly state anywhere that the grid-area for the `headerContent` is `chartHeader`?

File: charts/core/src/Chart/Chart.tsx [hunk 8]
```diff
@@ -45,12 +45,21 @@ export function Chart({
...
+             * but are used to provide a more declarative API for adding functionality
+             * to the chart canvas. They have access to the ChartContext and can be
+             * used to add components like Line, Grid, etc.
+             */}
+            {children}
+          </div>
           <div
             ref={chartRef}
-            className={`echart ${chartStyles}`}
+            className={cx('echart', chartStyles)}
```
Comment (shaneeza) [inline charts/core/src/Chart/Chart.tsx:60]: nit: why do you need `echart`?

File: charts/core/src/ChartCard/ChartCard.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  title: 'test',
+  headerContent: <div>{defaultContentText}</div>,
+  children: 'Chartchartchartchartchart',
+};
+
+const renderChartCard = (props?: Partial<ChartCardProps>) =>
+  render(<ChartCard {...defaultProps} {...props} />);
+
+const clickToggleButton = () => {
+  userEvent.click(screen.getByLabelText('Toggle button'));
```
Comment (shaneeza) [inline charts/core/src/ChartCard/ChartCard.spec.tsx:21]: nit: Thoughts on using a testid here? Also, will we be providing test utils for this component? If so, we have a pattern of creating a constant file to store testids. For example: https://github.com/mongodb/leafygreen-ui/blob/main/packages/select/src/constants.ts

File: charts/core/README.md [hunk 10]
```diff
@@ -58,66 +84,92 @@ Chart container component.
...
+
+Component for rendering a header in a chart.
+
+#### Props
+
+| Name                         | Description                                                                                                                                           | Type              | Default |
+| ---------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------- | ------- |
+| `title`                      | The title to display in the chart header.                                                                                                             | `string`          |         |
+| `showDivider` _(optional)_   | When true, renders a dividing line on top of header. Useful when stacking charts, such as in a `ChartGroup`.                                          | `boolean`         |         |
+| `headerContent` _(optional)_ | Content that will be rendered to the right of the title. Useful for adding components such as `IconButton`'s that control functionality in the chart. | `React.ReactNode` |         |
```
Comment (shaneeza) [inline charts/core/README.md:128]: nit: I didn't see a story that showcases this feature. Would you consider adding one to ensure that it works?

Comment (tsck) [reply charts/core/README.md:128]: Absolutely. I'm curious how we demo something like this since it takes in a `ReactNode`. Any suggestions?

File: charts/core/src/Chart/Chart.styles.ts (hunk 7 above)
Comment (tsck) [reply charts/core/src/Chart/Chart.styles.ts:6]: I'm not sure that's correct. `chartHeader` is the grid-area for `Header`, `headerContent` is a node that gets ultimately inside of the `Header`. Am I misunderstanding what's being asked here?

Comment (shaneeza) [reply charts/core/src/Chart/Chart.styles.ts:6]: My bad! I confused myself. Where did you add `grid-area: chartHeader`?

File: charts/core/src/ChartCard/ChartCard.spec.tsx (hunk 9 above)
Comment (shaneeza) [reply charts/core/src/ChartCard/ChartCard.spec.tsx:21]: Thanks for clarifying!

File: charts/core/README.md (hunk 10 above)
Comment (shaneeza) [reply charts/core/README.md:128]: Code might be a good example: https://storybook.mongodb.design/?path=/story/components-code--with-custom-actions.

I think you would need to have individual stories for standalone components.

File: charts/core/src/Chart/Chart.styles.ts (hunk 7 above)
Comment (tsck) [reply charts/core/src/Chart/Chart.styles.ts:6]: Great catch! Totally missed it. Since I happened to order it correctly in storybook I didn't realize I didn't actually add the area to the component. Will update!

File: charts/core/src/ChartCard/ChartCard.tsx [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+  leftInnerContainerStyles,
+  toggleButtonStyles,
+  toggleIconStyles,
+} from './ChartCard.styles';
+import { ChartCardProps } from './ChartCard.types';
+
+/**
+ * Card component that contains charts and can expand and collapse.
+ */
+export function ChartCard({
```
[the remark this answers is not in the document]
Comment (tsck) [reply charts/core/src/ChartCard/ChartCard.tsx:23]: I'm not sure but I can't think of a good reason for it to not? I can add it in the next push

File: charts/core/src/ChartCard/ChartCard.tsx (hunk 2 above)
Comment (tsck) [reply charts/core/src/ChartCard/ChartCard.tsx:70]: Gotcha, will change!

File: charts/core/src/ChartCard/ChartCard.tsx (hunk 5 above)
Comment (tsck) [reply charts/core/src/ChartCard/ChartCard.tsx:105]: This is actually an inline style and not a class, BUT it's leftover for something I was trying quickly inline that didn't work 😅 I'll clean this up

File: charts/core/src/Header/Header.styles.ts (hunk 6 above)
Comment (tsck) [reply charts/core/src/Header/Header.styles.ts:10]: Cool! Will do!

File: charts/core/README.md (hunk 10 above)
Comment (tsck) [reply charts/core/README.md:128]: I ended up doing a pretty big refactor here. I separated things into 2 stories "Chart" and "ChartCard", with the following stories:
- Chart
  - Basic
  - With Header Content 

- ChartCard
  - Basic
  - With Header Content
  - Containing Chart
  - Containing Multi Charts

I'd rather not do it in this PR (because folks are waiting to use this), but I actually think this is how the packages should ultimately be split as well. `@lg-charts/core` should probably really be `@lg-charts/chart` and only export what's needed for a chart itself. `@lg-charts/chart-card` should be separate and only export the card.

File: charts/core/src/ChartCard/ChartCard.styles.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  border: 1px solid
+    ${color[theme].border[Variant.Disabled][InteractionState.Default]};
+  border-radius: ${borderRadius[200]}px;
+  overflow: hidden;
+  width: 100%;
+  display: grid;
+  grid-template-rows: 40px 0fr;
+  transition: grid-template-rows ${transitionDuration.slower}ms ease-in-out;
+
+  &.open {
```
Comment (shaneeza) [inline charts/core/src/ChartCard/ChartCard.styles.ts:22]: Curious: why do you prefer using `.open` instead of conditionally rendering the styles with `cx`?

File: charts/core/README.md (hunk 10 above)
Comment (shaneeza) [reply charts/core/README.md:128]: sounds good to me

File: charts/core/src/ChartCard/ChartCard.styles.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (shaneeza) [reply charts/core/src/ChartCard/ChartCard.styles.ts]: i'm glad this worked!!!

File: charts/core/src/ChartCard/ChartCard.tsx [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+              className={toggleButtonStyles}
+              id={toggleId}
+              aria-label="Toggle button"
+              aria-controls={childrenId}
+              aria-expanded={isOpen}
+              onClick={handleToggleButtonClick}
+            >
+              <Icon
+                glyph="ChevronDown"
+                className={cx(toggleIconStyles, isOpen && 'open')}
```
Comment (shaneeza) [inline charts/core/src/ChartCard/ChartCard.tsx:82]: nit: 
```
[css`...`]:open
```

**Review by shaneeza — APPROVED** [review]

File: charts/core/src/ChartCard/ChartCard.styles.ts (hunk 12 above)
Comment (tsck) [reply charts/core/src/ChartCard/ChartCard.styles.ts:22]: Actually not a preference, just not used to having `cx` as an option! Will update this.
